core/src/test/java/org/apache/calcite/rex/RexBuilderTest.java

```diff
@@ -556,6 +556,28 @@ private void checkDate(RexNode node) {
     checkBigDecimalLiteral(builder, "-73786976294838206464");
   }
 
+  /** Tests {@link RexCopier#visitDynamicParam(RexDynamicParam)} */
+  @Test public void testCopyDynamicParam() {
+    final RelDataTypeFactory typeFactory = new SqlTypeFactoryImpl(RelDataTypeSystem.DEFAULT);
+    final RexBuilder builder = new RexBuilder(typeFactory);
+    final RelDataType type = typeFactory.createSqlType(SqlTypeName.VARCHAR);
+    final RexDynamicParam node = builder.makeDynamicParam(type, 1);
+    final RexDynamicParam copiedNode = (RexDynamicParam) builder.copy(node);
```

Comment:
I would put an additional assertion before performing the cast. That way the test will never fail with a ClassCastException which seems like a programming error but with an AssertionError which explicitly shows that the test is broken.